Container refactor fixes + pattern editor refresh

landonreed opened the pull request:
This PR addresses a few items discovered while working on some fixes for Pace Bus.

- fixes a few issues encountered with pattern editing (#395, #400, #399, fixes #397)
- fixes the editor lock overwrite functionality (also #395) and the redirect on login (#396)
- tweaks the layout of the pattern shape editor so that it's a bit more intuitive (see screengrab)
- replaces the direction icons for the pattern geometry with `leaflet-textpath` for better performance/legibility (see second screengrab)
- changes some components to `PureComponents` to prevent them from rendering as frequently (I'm not sure if this is effective in all the cases, so we may need to revisit some of the components, e.g. Sidebar).

![image](https://user-images.githubusercontent.com/2370911/51703939-bec40500-1fe5-11e9-887c-653d02a955f4.png)


![image](https://user-images.githubusercontent.com/2370911/51704040-fdf25600-1fe5-11e9-9eb5-2c98a6508daf.png)

evansiroky commented:
The CI runs are failing due to unresolved flow type errors.  Please fix those.

evansiroky commented on lib/editor/selectors/index.js:
```diff
@@ -125,6 +128,20 @@ const getPresentControlPoints = (state: AppState) => {
   return state.editor.editSettings.present.controlPoints
 }
 export const getTripCounts = (state: AppState) => state.editor.data.tables.trip_counts
+export const getActivePatternStops = createSelector(
```
Please write a unit test for this selector.

evansiroky commented on lib/editor/selectors/index.js:
```diff
@@ -360,10 +381,10 @@ function getControlPointsFromShapePoints (shapePoints: Array<ShapePoint>) {
   oldShapePoints: Array<ShapePoint>,
   patternStops: Array<PatternStop>,
-  stops: Array<any>, // TODO: use more exact type
+  activePatternStops: Array<GtfsStop>,
```
Would be fantastic to have a unit test for this function too.  Of course it'd be awesome to have unit tests for each possible scenario, but even a test that sent the oldShapePoints variable with a length being less than 2 is a good starting point as this big function has 0% coverage.

evansiroky commented on lib/editor/util/gtfs.js:
```diff
@@ -165,7 +165,7 @@ export const generateProps = (component: string, editorState: any): any => {
 }
 
-export function getEntityBounds (entity: any, offset: number = 0.005): ?latLngBounds {
+export function getEntityBounds (entity: any, stops?: Array<GtfsStop>, offset: number = 0.005): ?latLngBounds {
```
This could also use a very basic unit test.

evansiroky commented on lib/editor/util/map.js:
```diff
@@ -288,7 +288,7 @@ export async function recalculateShape ({
   newPoint?: LatLng,
   pattern: Pattern,
-  patternCoordinates: Coordinates,
+  patternCoordinates: Array<Coordinates>,
```
Another great candidate for adding tests.  Especially since there is already a file `lib/editor/util/__tests__/map.js_hold` with test cases for this exact function.

evansiroky commented:
The "hide inactive segments" checkbox could be greyed out unless there are some currently active segments.  Otherwise it's a little confusing since it's hiding everything all segments.

landonreed commented:
https://github.com/conveyal/datatools-ui/pull/401#issuecomment-458434242 is resolved with the [latest version of gtfs-lib](https://github.com/conveyal/gtfs-lib/releases/tag/v4.2.5).

landonreed commented:
@evansiroky, I have finally addressed your comments here, but this PR doesn't add unit tests. I think it's a higher priority to get this merged in though at this point to avoid any more dust gathering.

evansiroky commented:
https://github.com/conveyal/datatools-ui/pull/401#issuecomment-458434242 does not appear to be fixed while running datatools-server with gtfs-lib 4.2.5. Please see video posted in company slack.

evansiroky commented on lib/editor/util/__tests__/map.js:
```diff
@@ -56,23 +63,24 @@ describe('editor > util > map >', () => {
         editType: 'update',
         index: 0,
-        mapzenResponeFile: './fixtures/mapzen-response-update-first-stop.json',
+        // https://
```
please remove comment

evansiroky commented on lib/editor/util/__tests__/__snapshots__/map.js.snap:
```diff
@@ -0,0 +1,1244 @@
+        -77.922135,
+        42.089099,
+      ],
+    ],
```
The coordinates in this array don't go all the way to the new last stop. Is this the desired behavior?

This is what the coordinates look like before deleting the last stop:

<img width="332" alt="Screen Shot 2019-04-02 at 5 28 57 PM" src="https://user-images.githubusercontent.com/3112493/55444770-d6cb7e00-556c-11e9-854d-1f6212dc9229.png">

And here are those coordinates after:

<img width="362" alt="Screen Shot 2019-04-02 at 5 29 33 PM" src="https://user-images.githubusercontent.com/3112493/55444791-ecd93e80-556c-11e9-8af0-a4244db2f900.png">

evansiroky commented on lib/editor/util/__tests__/__snapshots__/map.js.snap:
```diff
@@ -0,0 +1,1244 @@
+      ],
+    ],
+  ],
+  "updatedControlPoints": Array [
```
This snapshot makes no sense due to bad input data. The pattern coordinates are in NY while the control point coordinates are in CA.

landonreed commented:
Yes, sorry about that. I was trying to adapt preexisting fixtures and that was a bad idea. I'll regenerate them from new input data.

evansiroky commented on lib/editor/util/__tests__/__snapshots__/map.js.snap:
```diff
@@ -834,96 +665,40 @@ Object {
-        -77.923104,
-        42.096481,
+        -77.92187333106995,
+        42.08646327816015,
```
This coordinate in the snapshot for `should update the first stop` seems out of place. This indicates that the shape goes backwards.

Here are the pattern coordinates and stops before the test:

<img width="348" alt="Screen Shot 2019-04-03 at 2 53 02 PM" src="https://user-images.githubusercontent.com/3112493/55516311-b6113000-5621-11e9-8646-32548a38b371.png">
 
And after recalcuate shape is ran:

<img width="356" alt="Screen Shot 2019-04-03 at 2 53 13 PM" src="https://user-images.githubusercontent.com/3112493/55516320-bd383e00-5621-11e9-8da6-0c17de066e3e.png">

The segment between the first and 2nd stop looks fine, but the shape doesn't continue to the last stop. Instead it goes backwards. You can't see if from that last picture, so here is a more zoomed in look:

<img width="653" alt="Screen Shot 2019-04-03 at 3 04 43 PM" src="https://user-images.githubusercontent.com/3112493/55516382-e1941a80-5621-11e9-9c69-1f9095c34152.png">

evansiroky commented on lib/editor/util/__tests__/__snapshots__/map.js.snap:
```diff
@@ -1030,76 +797,32 @@ Object {
-        -77.922203,
-        42.086407,
+        -77.9230535030365,
+        42.096176462786296,
```
The same odd thing in `should update the first stop` is now happening in reverse in the `should update the last stop`.

Before the test:

<img width="224" alt="Screen Shot 2019-04-03 at 3 11 40 PM" src="https://user-images.githubusercontent.com/3112493/55516800-36846080-5623-11e9-9c96-c879dcb642a4.png">

After:

<img width="239" alt="Screen Shot 2019-04-03 at 3 11 23 PM" src="https://user-images.githubusercontent.com/3112493/55516801-36846080-5623-11e9-9588-cdef1c534e60.png">

Zoomed in the first point is now going backwards to some other point and going forward.

<img width="556" alt="Screen Shot 2019-04-03 at 3 13 19 PM" src="https://user-images.githubusercontent.com/3112493/55516792-31271600-5623-11e9-8fde-e9f495f5d84f.png">

evansiroky commented on lib/editor/util/__tests__/fixtures/test-control-points.json:
```diff
@@ -40,16 +38,12 @@
   "distance": 905.1384059344052,
   "id": "0cj3",
-  "permanent": true,
+  "pointType": 2,
```
This should probably be `1` since there is no stopId for this control point.

evansiroky commented on lib/editor/util/__tests__/__snapshots__/map.js.snap:
```diff
@@ -583,46 +448,22 @@ Object {
-        -77.922173,
-        42.086498,
+        -77.9230535030365,
+        42.096176462786296,
```
This snapshot for `should update a point in the middle of the control points` seems really off again.

Before:

<img width="235" alt="Screen Shot 2019-04-03 at 4 41 44 PM" src="https://user-images.githubusercontent.com/3112493/55520076-8406ca80-562f-11e9-9ce4-0e33bf3048c2.png">

After:

<img width="239" alt="Screen Shot 2019-04-03 at 4 41 35 PM" src="https://user-images.githubusercontent.com/3112493/55520087-8a954200-562f-11e9-9892-c9526a403b24.png">

evansiroky commented on lib/editor/util/__tests__/__snapshots__/map.js.snap:
```diff
@@ -160,70 +125,30 @@ Object {
     Array [
       Array [
-        -77.922135,
-        42.089099,
```
The result of the `should delete the first stop` test implies that the last stop is deleted and the line is backwards!

Before:

<img width="241" alt="Screen Shot 2019-04-03 at 5 06 01 PM" src="https://user-images.githubusercontent.com/3112493/55520818-d1386b80-5632-11e9-8650-36ebc2052533.png">

After:

<img width="236" alt="Screen Shot 2019-04-03 at 5 05 52 PM" src="https://user-images.githubusercontent.com/3112493/55520816-ced61180-5632-11e9-8a6a-d134766a0471.png">

landonreed commented:
Agh, you're right. Sorry about that. I'm going to add some log statements that actually print out geojson.io links to make this stuff a little easier to debug.

landonreed commented:
:tada: This PR is included in version 4.1.0 :tada:

The release is available on [GitHub release](https://github.com/ibi-group/datatools-ui/releases/tag/v4.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
